extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/DataSourcesJdbcBuildTimeConfig.java

```diff
@@ -5,35 +5,35 @@
 import io.quarkus.runtime.annotations.ConfigDocMapKey;
 import io.quarkus.runtime.annotations.ConfigDocSection;
 import io.quarkus.runtime.annotations.ConfigGroup;
-import io.quarkus.runtime.annotations.ConfigItem;
 import io.quarkus.runtime.annotations.ConfigPhase;
 import io.quarkus.runtime.annotations.ConfigRoot;
+import io.smallrye.config.ConfigMapping;
+import io.smallrye.config.WithParentName;
 
-@ConfigRoot(name = "datasource", phase = ConfigPhase.BUILD_AND_RUN_TIME_FIXED)
-public class DataSourcesJdbcBuildTimeConfig {
+@ConfigMapping(prefix = "quarkus.datasource")
+@ConfigRoot(phase = ConfigPhase.BUILD_AND_RUN_TIME_FIXED)
+public interface DataSourcesJdbcBuildTimeConfig {
 
     /**
      * The default datasource.
      */
-    @ConfigItem
-    public DataSourceJdbcBuildTimeConfig jdbc;
+    DataSourceJdbcBuildTimeConfig jdbc();
 
     /**
      * Additional named datasources.
      */
     @ConfigDocSection
     @ConfigDocMapKey("datasource-name")
-    @ConfigItem(name = ConfigItem.PARENT)
-    public Map<String, DataSourceJdbcOuterNamedBuildTimeConfig> namedDataSources;
+    @WithParentName
+    Map<String, DataSourceJdbcOuterNamedBuildTimeConfig> namedDataSources();
```

Comment:
The idea was to add support for default values in a Map. I started some prototypes and experimented with some ideas,  and I think I had a plan to implement that, but it is not done yet.

Adding a specific class to support this, while possible, of course, would fall off the traditional way of mapping stuff. Until now, we were able to use standard types to represent pieces of the configuration. Adding specific constructs will always require a lot of adoption effort (and probably some frustration on why can't you use what you already have).

What will not be inline with a regular `Map`? I'm thinking that a default will not be listed in `.keySet()`, `entrySet()`, and `containsKey()`, but a `get` will return you a default if there is no value for that key.

Reply:
Also, unofficially you can try this to initialize a group programmatically. I haven't tried that with complex examples, but it should work fine, as long as every mapping has a default defined or you get an error:

```java
SmallRyeConfig config = ConfigProvider.getConfig().unwrap(SmallRyeConfig.class);
ConfigMappingContext context = new ConfigMappingContext(config);
// you need to add the path to the map, plus `*` to represent the key. This will not be added to the main mapping `Map`
context.getStringBuilder().replace(0, context.getStringBuilder().length(), "map.nested.*");
ConfigGroup group = context.constructGroup(ConfigGroup.class);
```

Reply:
> What will not be inline with a regular Map?

> What will not be inline with a regular `Map`? I'm thinking that a default will not be listed in `.keySet()`, `entrySet()`, and `containsKey()`, but a `get` will return you a default if there is no value for that key.

Yes, that.

Also, and perhaps more importantly, unless you make it configurable through annotations this is a breaking change. There's a decent chance some code out there (probably even in Quarkus) was using `map.get("something") == null` to test whether something was configured or not. Yes I know there is `containsKey`, but `var val = config.get(...); if ( val == null ) {...} else {...}` is a common pattern when you need to use the value if it's there.

By hijacking `Map` to support more than its contract, we would break such existing code.

> Adding specific constructs will always require a lot of adoption effort

Note we're not requesting that anyone use that theoretical new type. It's not a replacement for existing features, it comes in addition to existing `Map` support. I'm not sure how adoption effort would be relevant in that case?

> (and probably some frustration on why can't you use what you already have).

And implementing behavior that does not match the `Map` contract (see above) will lead to nasty bugs, which is *also* frustrating.

Even if the behavior is configurable through annotations, I wouldn't be surprise to see it bite someone who comes back to the code after a while and just expects a `Map` to behave like a `Map`.

> Also, unofficially you can try this to initialize a group programmatically

Nice trick, maybe we could just add the feature to Quarkus then? Then we have fewer constraints as to how it should be exposed (custom type vs. `Map`).